src/Symfony/Component/DependencyInjection/ParameterBag/ParameterResolver.php

```diff
@@ -0,0 +1,159 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\ParameterBag;
+
+use Symfony\Component\DependencyInjection\Exception\ParameterNotFoundException;
+use Symfony\Component\DependencyInjection\Exception\ParameterCircularReferenceException;
+use Symfony\Component\DependencyInjection\Exception\RuntimeException;
+
+/**
+ * Replaces parameter placeholders (%name%) by their values.
+ *
+ * @author Ener-Getick <egetick@gmail.com>
+ */
+final class ParameterResolver
+{
+    /**
+     * @var callable
+     */
+    private $parameterGetter;
+
+    /**
+     * @var bool
+     */
+    private $resolved;
+
+    /**
+     * Constructor.
+     *
+     * @param callable $parameterGetter
+     * @param bool     $resolved
+     */
+    public function __construct(callable $parameterGetter, $resolved = false)
+    {
+        $this->parameterGetter = $parameterGetter;
+        $this->resolved = $resolved;
+    }
+
+    /**
+     * Replaces parameter placeholders (%name%) by their values.
+     *
+     * @param mixed $value     A value
+     * @param array $resolving An array of keys that are being resolved (used internally to detect circular references)
+     *
+     * @return mixed The resolved value
+     *
+     * @throws ParameterNotFoundException          if a placeholder references a parameter that does not exist
+     * @throws ParameterCircularReferenceException if a circular reference is detected
+     * @throws RuntimeException                    when a given parameter has a type problem.
+     */
+    public function resolveValue($value, array $resolving = array())
```

Comment:
I think we shouldn't expose the `$resolving` argument. The public interface should only expect the `$value` and hide everything else in private methods.